chemprop/uncertainty/uncertainty_calibrator.py

```diff
@@ -615,6 +621,265 @@ def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
         return uncal_preds.tolist(), cal_preds.tolist()
 
 
+class ConformalMulticlassCalibrator(UncertaintyCalibrator):
+    """
+    Conformal Calibrator.
+    """
+    @property
+    def label(self):
+        return f"{self.uncertainty_method}_conformal_multiclass"
+
+    def raise_argument_errors(self):
+        super().raise_argument_errors()
+        if self.dataset_type != "multiclass":
+            raise ValueError(
+                "Conformal is only implemented for multiclass dataset types."
+            )
+
+    def nonconformity_score(self, i, j, task_id, uncal_preds):#x,y are the indices
+        return -uncal_preds[i][task_id][j]
+
+    def calibrate(self):
```

Comment:
There is a more concise and faster way to write these functions, while avoiding long for loops over all the data. I have written some quick but untested code that does what you want to do this way.

```python
@staticmethod
def nonconformity_scores(preds):
    """Fixed per class. Example is for adaptive.                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                              
    Args:                                                                                                                                                                                                                                                                     
        preds: [num_examples, num_tasks, num_classes]                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                                              
    Returns:                                                                                                                                                                                                                                                                  
        scores: [num_examples, num_tasks, num_classes]                                                                                                                                                                                                                        
    """
    # Sort from high to low.                                                                                                                                                                                                                                                  
    sort_inds = np.argsort(-preds, axis=-1)

    # Sort, and take cumulative sums.                                                                                                                                                                                                                                         
    sorted_scores = preds[sort_inds].cumsum(axis=-1)

    # Map back to original order.                                                                                                                                                                                                                                             
    unsort_inds = np.argsort(sort_inds, axis=-1)
    unsorted_scores = sorted_scores[unsorted_inds]

    return unsorted_scores

def calibrate(self):
    # shape(num_examples, num_tasks, num_classes)                                                                                                                                                                                                                             
    uncal_preds = np.array(self.calibration_predictor.get_uncal_preds())

    # shape(num_examples, num_tasks)                                                                                                                                                                                                                                          
    targets = np.array(self.calibration_data.targets(), dtype=int)

    # Keep mask of invalid targets (missing data). Replace Nans with 0.                                                                                                                                                                                                       
    mask = np.isnan(targets)
    targets = np.nan_to_num(targets, copy=True, nan=0, posinf=None, neginf=None)

    # Compute all non-conformity scores for all examples/tasks/classes.                                                                                                                                                                                                       
    all_scores = self.nonconformity_scores(uncal_preds)

    # Take scores of true targets only.                                                                                                                                                                                                                                       
    target_scores = np.take_along_axis(all_scores, np.expand_dims(targets, -1), -1).squeeze(-1)

    # For every task, compute calibration set quantile.                                                                                                                                                                                                                       
    self.qhats = []
    num_tasks = targets.shape[1]
    for task_id in range(num_tasks):
        # Get all vaid non-conformity scores for this task.                                                                                                                                                                                                                   
        task_scores = target_scores[mask[:, task_id], task_id].reshape(-1)

        # Pad with +inf.                                                                                                                                                                                                                                                      
        task_scores = np.pad(task_scores, (0, 1), "constant", constant_values=np.Inf)

        # Compute empirical quantile.                                                                                                                                                                                                                                         
        qhat = np.quantile(task_scores, 1 - self.alpha / num_tasks, interpolation="upper")
        qhats.append(qhat)


def apply_calibration(self, uncal_predictor):
    # shape(num_examples, num_tasks, num_classes)                                                                                                                                                                                                                             
    uncal_preds = np.array(uncal_predictor.get_uncal_preds())  # shape(data, task)                                                                                                                                                                                            

    # 0/1 for if candidate class y is included in output set.                                                                                                                                                                                                                 
    # shape(num_examples, num_tasks, num_classes)                                                                                                                                                                                                                             
    cal_preds = np.zeros_like(uncal_preds)
    for task_id, qhat in enumerate(self.qhats):
        cal_preds[:, task_id, :] = uncal_preds[:, task_id, :] <= qhat

    return uncal_preds.tolist(), cal_preds.tolist()
```